src/USER-MDI/fix_mdi_engine.cpp

```diff
@@ -0,0 +1,932 @@
+
+  // convert the energy to atomic units
+  pe *= kelvin_to_hartree/force->boltz;
+
+  if (master) {
+    ierr = MDI_Send((char*) send_energy, 1, MDI_DOUBLE, driver_socket);
+    if (ierr != 0)
+      error->all(FLERR,"Unable to send potential energy to driver");
+  }
+}
+
+
+void FixMDIEngine::send_ke(Error* error)
+{
+  double kelvin_to_hartree;
+  MDI_Conversion_Factor("kelvin_energy", "hartree", &kelvin_to_hartree);
+
+  double ke = kinetic_energy;
+  double *send_energy = &ke;
+
+  // convert the energy to atomic units
+  ke *= kelvin_to_hartree/force->boltz;
+
+  if (master) {
+    ierr = MDI_Send((char*) send_energy, 1, MDI_DOUBLE, driver_socket);
+    if (ierr != 0)
+      error->all(FLERR,"Unable to send potential energy to driver");
+  }
+}
+
+
+void FixMDIEngine::send_types(Error* error)
+{
+  int * const type = atom->type;
+
+  if (master) { 
+    ierr = MDI_Send((char*) type, atom->natoms, MDI_INT, driver_socket);
+    if (ierr != 0)
+      error->all(FLERR,"Unable to send atom types to driver");
+  }
+}
+
+
+void FixMDIEngine::send_labels(Error* error)
+{
+  char *labels = new char[atom->natoms * MDI_LABEL_LENGTH];
+  memset(labels, ' ', atom->natoms * MDI_LABEL_LENGTH);
+
+  for (int iatom=0; iatom < atom->natoms; iatom++) {
+    std::string label = std::to_string( atom->type[iatom] );
+    int label_len = std::min( int(label.length()), MDI_LABEL_LENGTH );
+    strncpy(&labels[iatom * MDI_LABEL_LENGTH], label.c_str(), label_len);
+  }
+
+  if (master) { 
+    ierr = MDI_Send( labels, atom->natoms * MDI_LABEL_LENGTH, MDI_CHAR, driver_socket);
+    if (ierr != 0)
+      error->all(FLERR,"Unable to send atom types to driver");
+  }
+
+  delete [] labels;
+}
+
+
+void FixMDIEngine::send_masses(Error* error)
+{
+  double * const mass = atom->mass;
+  int * const type = atom->type;
+
+  if (master) { 
+    double *mass_by_atom = new double[atom->natoms];
+    for (int iatom=0; iatom < atom->natoms; iatom++) {
+      mass_by_atom[ atom->tag[iatom] - 1 ] = mass[ type[iatom] ];
+    }
+    ierr = MDI_Send((char*) mass_by_atom, atom->natoms, MDI_DOUBLE, driver_socket);
+    if (ierr != 0)
+      error->all(FLERR,"Unable to send atom masses to driver");
+    delete [] mass_by_atom;
+  }
+
+}
+
+
+void FixMDIEngine::send_forces(Error* error)
+{
+  double angstrom_to_bohr;
+  MDI_Conversion_Factor("angstrom", "bohr", &angstrom_to_bohr);
+  double kelvin_to_hartree;
+  MDI_Conversion_Factor("kelvin_energy", "hartree", &kelvin_to_hartree);
+
+  double potconv, posconv, forceconv;
+  potconv=kelvin_to_hartree/force->boltz;
+  posconv=force->angstrom/angstrom_to_bohr;
+  forceconv=potconv*posconv;
+
+  double *forces;
+  double *forces_reduced;
+  double *x_buf;
+
+  int *mask = atom->mask;
+  int nlocal = atom->nlocal;
+
+  forces = new double[3*atom->natoms];
+  forces_reduced = new double[3*atom->natoms];
+  x_buf = new double[3*nlocal];
+
+  // zero the forces array
+  for (int iforce = 0; iforce < 3*atom->natoms; iforce++) {
```

Comment:
This will overflow for large systems. LAMMPS can handle systems larger than 2 billion atoms: https://docs.lammps.org/Build_settings.html#size
This specific loop will overflow with 716M atoms already.

Also, in general, it can negatively impact performance to allocate and free buffers in every call.